src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionManager.cs

```diff
@@ -1,41 +1,60 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
 using System.Threading.Tasks;
 using Microsoft.Azure.SignalR.Protocol;
 
 namespace Microsoft.Azure.SignalR.AspNet
 {
     internal class ServiceConnectionManager : IServiceConnectionManager
     {
-        public void AddConnection(string hubName, IServiceConnection connection)
+        private readonly ConcurrentDictionary<string, IServiceConnectionContainer> _serviceConnections = new ConcurrentDictionary<string, IServiceConnectionContainer>();
+                
+        public void AddConnection(string hubName, IServiceConnectionContainer connection)
         {
-            throw new System.NotImplementedException();
+            _serviceConnections.TryAdd(hubName, connection);
         }
 
         public Task StartAsync()
         {
-            throw new System.NotImplementedException();
+            return Task.WhenAll(GetConnections().Select(s => s.StartAsync()));
         }
 
         public Task StopAsync()
         {
-            throw new System.NotImplementedException();
+            return Task.WhenAll(GetConnections().Select(s => s.StopAsync()));
         }
 
-        public IServiceConnection WithHub(string hubName)
+        public IServiceConnectionContainer WithHub(string hubName)
         {
-            throw new System.NotImplementedException();
+            if (!_serviceConnections.TryGetValue(hubName, out var connection))
+            {
+                throw new KeyNotFoundException($"Service connection with Hub {hubName} does not exist");
+            }
+
+            return connection;
+        }
+
+        public Task WriteAsync(string partitionKey, ServiceMessage serviceMessage)
+        {
+            return Task.WhenAll(GetConnections().Select(s => s.WriteAsync(partitionKey, serviceMessage)));
         }
 
         public Task WriteAsync(ServiceMessage serviceMessage)
         {
-            throw new System.NotImplementedException();
+            return Task.WhenAll(GetConnections().Select(s => s.WriteAsync(serviceMessage)));
         }
 
-        public Task WriteAsync(string connectionId, object message)
+        private IEnumerable<IServiceConnectionContainer> GetConnections()
         {
-            throw new System.NotImplementedException();
+            foreach(var pair in _serviceConnections)
```

Comment:
use linq?

Reply:
This is a preparation review for an upcoming one, this method is preserved for more changes. 😄